File: Commons/Reflection/src/main/java/com/greentree/commons/reflection/info/TypeUtil.java
package com.greentree.commons.reflection.info;

import com.greentree.commons.util.iterator.IteratorUtil;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;

public class TypeUtil {

    private static final TypeInfo<Object> OBJECT_TYPE = TypeInfoBuilder.getTypeInfo(Object.class);

    public static boolean isExtends(Class<?> superType, TypeInfo<?> type) {
        return isExtends(TypeInfoBuilder.getTypeInfo(superType), type);
    }

    @SuppressWarnings("unchecked")
    public static <T> boolean isExtends(TypeInfo<T> superType, TypeInfo<?> type) {
        if (superType.equals(type))
            return true;
        return superType.isSuperTo((TypeInfo<? extends T>) type);
    }

    public static boolean isExtends(TypeInfo<?> superType, Class<?> type) {
        return isExtends(superType, TypeInfoBuilder.getTypeInfo(type));
    }

    public static boolean isExtends(Class<?> superType, Class<?> type) {
        return isExtends(TypeInfoBuilder.getTypeInfo(superType), TypeInfoBuilder.getTypeInfo(type));
    }

    public static <T> TypeInfo<?> lca(Iterable<? extends TypeInfo<? extends T>> iter) {
        return IteratorUtil.reduce(iter, (a, b) -> lca(a, b), () -> OBJECT_TYPE);
    }

    public static <T> TypeInfo<?> lca(TypeInfo<? extends T> a, TypeInfo<? extends T> b) {
        if (isExtends(a, b))
            return a;
        if (isExtends(b, a))
            return b;
        final var lcas = new HashSet<TypeInfo<?>>();
        for (var i : getSuperClassAndInterfaces(a)) {
            final var lca = lca(i, b);
            lcas.add(lca);
        }
        for (var i : getSuperClassAndInterfaces(b)) {
            final var lca = lca(a, i);
            lcas.add(lca);
        }
        lcas.remove(OBJECT_TYPE);
        if (lcas.isEmpty())
            return OBJECT_TYPE;
        if (lcas.size() == 1)
            return lcas.iterator().next();
        for (var lca : lcas)
            if (!lca.isInterface())
                return lca;
        throw new IllegalArgumentException("lcas " + lcas + " of " + a + " " + b);
    }

    public static <T, S> TypeInfo<T> getFirstArgument(Class<? extends S> type,
                                                      Class<S> superClass) {
        return getFirstArgument(TypeInfoBuilder.getTypeInfo(type), superClass);
    }

    @SuppressWarnings("unchecked")
    public static <T, S> TypeInfo<T> getFirstArgument(TypeInfo<? extends S> type,
                                                      Class<S> superClass) {
        return (TypeInfo<T>) getType(type, superClass).getTypeArguments()[0];
    }

    @SuppressWarnings("unchecked")
    public static <T> TypeInfo<T> getType(TypeInfo<? extends T> type, Class<T> superClass) {
        if (type.toClass().equals(superClass))
            return (TypeInfo<T>) type;
        final var superType = TypeInfoBuilder.getTypeInfo(superClass);
        for (var i : getSuperClassAndInterfaces(type))
            if (isExtends(superType, i)) {
                final TypeInfo<T> t = getType((TypeInfo<? extends T>) i, superClass);
                if (t != null)
                    return t;
            }
        return null;
    }

    public static <T> Iterable<TypeInfo<? super T>> getSuperClassAndInterfaces(TypeInfo<T> info) {
        final Collection<TypeInfo<? super T>> res = new ArrayList<>();
        final var s = info.getSuperType();
        if (s != null)
            res.add(s);
        Collections.addAll(res, info.getInterfaces());
        return res;
    }

}

File: Commons/Reflection/src/main/java/com/greentree/commons/reflection/info/TypeUtil.kt
package com.greentree.commons.reflection.info

import com.greentree.commons.reflection.info.TypeInfoBuilder.getTypeInfo
import kotlin.reflect.KClass

object TypeUtil {

	@JvmStatic
	fun isExtends(superType: Class<*>, type: Class<*>): Boolean {
		return isExtends(getTypeInfo(superType), getTypeInfo(type))
	}

	@JvmStatic
	fun isExtends(superType: Class<*>, type: KClass<*>): Boolean {
		return isExtends(getTypeInfo(superType), getTypeInfo(type))
	}

	@JvmStatic
	fun isExtends(superType: Class<*>, type: TypeInfo<*>): Boolean {
		return isExtends(getTypeInfo(superType), type)
	}

	@JvmStatic
	fun isExtends(superType: KClass<*>, type: Class<*>): Boolean {
		return isExtends(getTypeInfo(superType), getTypeInfo(type))
	}

	@JvmStatic
	fun isExtends(superType: KClass<*>, type: KClass<*>): Boolean {
		return isExtends(getTypeInfo(superType), getTypeInfo(type))
	}

	@JvmStatic
	fun isExtends(superType: KClass<*>, type: TypeInfo<*>): Boolean {
		return isExtends(getTypeInfo(superType), type)
	}

	@JvmStatic
	fun isExtends(superType: TypeInfo<*>, type: Class<*>): Boolean {
		return isExtends(superType, getTypeInfo(type))
	}

	@JvmStatic
	fun isExtends(superType: TypeInfo<*>, type: KClass<*>): Boolean {
		return isExtends(superType, getTypeInfo(type))
	}

	@JvmStatic
	fun isExtends(superType: TypeInfo<*>, type: TypeInfo<*>): Boolean {
		if(superType == type)
			return true
		return superType.isParentFor(type)
	}

	@JvmStatic
	fun <S> getSuperType(type: TypeInfo<out S>, superClass: Class<S>): TypeInfo<S> =
		type.getSuperType(superClass) as TypeInfo<S>

	@JvmStatic
	fun <S, T> getFirstArgument(
		type: TypeInfo<out S>,
		superClass: Class<S>,
	): Class<out T> = getSuperType(type, superClass).typeArguments[0].toClass() as Class<out T>

	@JvmStatic
	fun <S : Any, T> getFirstArgument(
		type: TypeInfo<out S>,
		superClass: KClass<S>,
	): Class<out T> = getFirstArgument(type, superClass.java)

	@JvmStatic
	fun <S : Any, T> getFirstArgument(
		type: KClass<out S>,
		superClass: Class<S>,
	): Class<out T> = getFirstArgument(getTypeInfo(type), superClass)

	@JvmStatic
	fun <S : Any, T> getFirstArgument(
		type: KClass<out S>,
		superClass: KClass<S>,
	): Class<out T> = getFirstArgument(getTypeInfo(type), superClass.java)

	@JvmStatic
	fun <S : Any, T> getFirstArgument(
		type: Class<out S>,
		superClass: Class<S>,
	): Class<out T> = getFirstArgument(getTypeInfo(type), superClass)

	@JvmStatic
	fun <S : Any, T> getFirstArgument(
		type: Class<out S>,
		superClass: KClass<S>,
	): Class<out T> = getFirstArgument(getTypeInfo(type), superClass.java)

	fun <T> getSuperClassAndInterfacesAsClass(cls: Class<T>) = sequence {
		cls.superclass?.let { superClass ->
			yield(superClass)
		}
		for(i in cls.interfaces) {
			yield(i)
		}
	}
}
